google-cloud-pubsub/lib/google/cloud/pubsub/flow_controller.rb

```diff
@@ -0,0 +1,159 @@
+
+        def initialize message_limit: 1000, byte_limit: 10_000_000, limit_exceeded_behavior: :ignore
+          # init MonitorMixin
+          # super()
+          unless [:ignore, :error, :block].include? limit_exceeded_behavior
+            raise ArgumentError, "limit_exceeded_behavior must be one of :ignore, :error, :block"
+          end
+          @mutex = Mutex.new
+          @message_limit = message_limit
+          @byte_limit = byte_limit
+          @limit_exceeded_behavior = limit_exceeded_behavior
+          @outstanding_messages = 0
+          @outstanding_bytes = 0
+
+          @awaiting_message_acquires = []
+          @awaiting_bytes_acquires = []
+        end
+
+        def acquire message_size
+          return if limit_exceeded_behavior == :ignore
+          @mutex.lock
+          if limit_exceeded_behavior == :error && @outstanding_messages + 1 > message_limit
+            raise FlowControlLimitError, "Flow control message limit (#{message_limit}) would be exceeded"
+          end
+          if limit_exceeded_behavior == :error && @outstanding_bytes + message_size > byte_limit
+            raise FlowControlLimitError,
+                  "Flow control byte limit (#{byte_limit}) would be exceeded, message_size: #{message_size}"
+          end
+          if limit_exceeded_behavior == :block && message_limit < 1
+            raise FlowControlLimitError,
+                  "Flow control message limit (#{message_limit}) exceeded by a single message, would block forever"
+          end
+          if limit_exceeded_behavior == :block && message_size > byte_limit
+            raise FlowControlLimitError,
+                  "Flow control byte limit (#{byte_limit}) exceeded by a single message, would block forever"
+          end
+
+          acquire_message
+          acquire_bytes message_size
+        ensure
+          @mutex.unlock if @mutex.locked?
+        end
+
+        # rubocop:disable Style/IdenticalConditionalBranches
+        # rubocop:disable Style/GuardClause
+
+        def acquire_message
+          waiter = nil
+          while @outstanding_messages + 1 > message_limit
+            if waiter.nil?
+              waiter = Concurrent::Event.new
+              # This waiter gets added to the back of the line.
+              @awaiting_message_acquires << waiter
+            else
+              waiter = Concurrent::Event.new
+              # This waiter already in line stays at the head of the line.
+              @awaiting_message_acquires[0] = waiter
+            end
+            @mutex.unlock
+            waiter.wait
+            @mutex.lock
+          end
+          @outstanding_messages += 1
+
+          @awaiting_message_acquires.shift if waiter # Remove the newly released waiter from the head of the queue.
+
+          # There may be some surplus messages left; let the next message waiting for a token have one.
+          if !@awaiting_message_acquires.empty? && @outstanding_messages < message_limit
```

Comment:
This is probably not necessary for the message counter, since it decrements only by 1 each time, and every time `release` will ping the first semaphore. (However, if we combine the queues, then yes, we will still need this logic for the combined conditional.)

Reply:
If we stick with the separate queues, then yes, I agree that we can probably remove this.